source/generators/MuonAngleGenerator.cc

```diff
@@ -175,20 +245,98 @@ G4String MuonAngleGenerator::MuonCharge() const
 
 void MuonAngleGenerator::GetDirection(G4ThreeVector& dir)
 {
-  // GetAngles from file?? Azimuth defined anticlockwise
-  // From north
-  G4double zenith  = 0.;
-  G4double azimuth = 0.;
-  distribution_->GetRandom2(azimuth, zenith);
-  // !! Current distribution in units of pi
-  zenith  *= pi;
-  azimuth *= pi;
-
-  dir.setX(sin(zenith) * sin(azimuth));
-  dir.setY(-cos(zenith));
-  dir.setZ(-sin(zenith) * cos(azimuth));
-
-  dir *= *rPhi_;
+
+  // Bool to check if zenith has a valid value. If not then resample
+  G4bool invalid_evt = true;
+
+  while(invalid_evt){
+  
+    // Amount to smear the randomly sampled zenith/azimuth values by
+    G4double zen_BW_smear{1.0e6}; 
+    G4double az_BW_smear{1.0e6};
+
+    // Generate random index weighted by the bin contents
+    G4int RN_indx = discr_dist_(RN_engine_);
+
+    // Loop over the zenith values and find the corresponding bin width to smear
+    for (G4int i = 0; i < zenith_bins_.size()-1; i++){
+        
+        // Catch very last bin
+        if (zen_BW_smear == 1e6 && i == zenith_bins_.size()-2){
+          if (zeniths_[RN_indx] >= zenith_bins_[i] 
+              && zeniths_[RN_indx] <= zenith_bins_[i+1]){
+            
+              zen_BW_smear = zen_BW_[i];
+            }
+        }
+        else {
+          if (zeniths_[RN_indx] >= zenith_bins_[i] 
+                && zeniths_[RN_indx] < zenith_bins_[i+1]){
+              
+              zen_BW_smear = zen_BW_[i];
+
+          }
+        }
+    
+    }
+
+    // Loop over the azimuth values and find the corresponding bin width to smear
+    for (G4int i = 0; i < azimuth_bins_.size()-1; i++){
+        
+        // Include last bin in check
+        if (az_BW_smear == 1e6 && i == azimuth_bins_.size()-2){
+          if (azimuths_[RN_indx] >= azimuth_bins_[i] 
+              && azimuths_[RN_indx] <= azimuth_bins_[i+1]){
+            
+              az_BW_smear = az_BW_[i];
+
+          }
+        }
+        else {
+
+          if (azimuths_[RN_indx] >= azimuth_bins_[i] 
+                && azimuths_[RN_indx] < azimuth_bins_[i+1]){
+              
+              az_BW_smear = az_BW_[i];
+
+          }
+        }
+
+    }
+
+    // Check if the smear values are set properly
+    if (az_BW_smear == 1.0e6 || zen_BW_smear == 1.0e6 )
+      G4Exception("[MuonAngleGenerator]", "GetDirection()",
+                FatalException, " bin smear value was not set correctly ");
+
+    // Gaussian dist to smear by bin widths in azimuth and zenith
+    std::normal_distribution<G4double> Gauss_az(0, az_BW_smear);
+    std::normal_distribution<G4double> Gauss_zen(0, zen_BW_smear);
+    
+    // Get the Gaussian smear values 
+    G4double az_smear  = Gauss_az(RN_engine_az_);
```

Comment:
As commented above, I would use G4 builtins, in this case `G4RandGauss::shoot`

Reply:
Updated generator to use the CLHEP RandGauss tools rather than the std library.

Reply:
I would avoid defining two random engines but use `G4RandGauss::shoot` instead, what do you think @jmalbos @paolafer?